### DISCUSSION TITLE
Can't modify chat_data when using job queue

### DISCUSSION THREAD

**Discussion by avfawkes** [python-telegram-bot/python-telegram-bot#3681] [Q&A]
I have a task to notify some bot users about today birthdays among of all bot users. So in main() i start the job:
```
    application.job_queue.run_daily(callback=today_bd_notifier,
                                    time=time.fromisoformat('09:00:00+04:00'),
                                    days=(0, 1, 2, 3, 4, 5, 6),
                                    data=application.chat_data,
                                    name='today_bd_notifier_job')

```

Callback looks like:
```
async def today_bd_notifier(context: CallbackContext) -> None:
    # here i'm omitting some code where i search the database for birthday people, then send messages to those who need to be informed and try to store some information in the chat_data of those users
    for t_user_id in recepients:
        context.job.data[t_user_id]['pointer'] = 1
        context.job.data[t_user_id]['search_results'] = birthdays_df
        context.job.data[t_user_id]['last_message'] = await context.bot.send_message(chat_id=t_user_id, text=text, parse_mode=ParseMode.HTML, reply_markup=context.bot_data['verbose_button'])

```

Next step is a click of inlineButton "verbose_button" It should show detailed information about birthday person, which i expect to be in "context.chat_data"
```
async def button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    query = update.callback_query
    pointer = context.chat_data['pointer']
    search_results = context.chat_data['search_results']
    .......
```

But "context.chat_data" is empty.

I supposed "application.chat_data" is a pointer to dict and i can change it. In fact it just local variable. How can i change the "chat_data" from callback of job queue task?

**Comment by Poolitzer** [comment]
Jobs are not connected to a chat unless you set chat_id when creating it.

**Reply by Bibo-Joshi** [reply]
he's writing something in there though: `context.job.data[t_user_id]['search_results'] = birthdays_df`

**Comment by Bibo-Joshi** [comment]
Some comments:

1. there is no need to pass `data=application.chat_data` - you cann just access that via `context.application.chat_data`
2. When you do `context.job.data[t_user_id]['search_results'] = birthdays_df` - is the `birthdays_df` the birthday of the user associated with `t_user_id`? Or whos birthday is it? Because in the callback of `CallbackQueryHandler`, `chat_data` will belong to the chat where the button is pressed in
3. Since you are talking about users rather than chats, `user_data` might be more suitable than `chat_data`
4. In general I recommend to store data in `bot_data` if it should be available for different chats
5.  > I supposed "application.chat_data" is a pointer to dict and i can change it

    That half is correct. See the docs of [`Application.chat_data`](https://docs.python-telegram-bot.org/en/stable/telegram.ext.application.html#telegram.ext.Application.chat_data) for more details.

**Reply by avfawkes** [reply]
2. `birthdays_df` is Pandas DataFrame of users, whose birthday today. `t_user_id` is telegram user id of person who need to be informed about today birthdays. So, for example its 100 person in organization and few of them daily need to be informed about birthdays
3.  In case of user chat with bot there is no functional difference, except how it named, if i understand right
4. Yes, but its already a lot of code, where used `chat_data` as data source, so i try to reuse it without modification, by passing `birthday_df` from job callback to some chats `chat_data`
5. Expected to be, but not work.

Let me try to explain in simple abstract example:
```
application.job_queue.run_daily(callback, time)

async def callback(context: CallbackContext) -> None:
    chat_id = '12345678'   # person who need to be informed
    text = 'today we wish happy birthday to the following colleagues....'
    context.application.chat_data[chat_id] = {1: 'some text', 2: 'another text'}
    await context.bot.send_message(chat_id=chat_id, text=text, reply_markup=InlineKeyboardMarkup([[InlineKeyboardButton('verbose', callback_data='verbose')]])

async def verbose_button_callback(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    update.callback_query.answer()
    data = context.chat_data
    .....
```
Here I expect `data` to be  `{1: 'some text', 2: 'another text'}`, but its empty. What's wrong?

**Reply by Bibo-Joshi** [reply]
Point 3: TG recently emphasized that `user_id` and `chat_id` need not coincide. While they currently still do, I personally do see this as occasion to make the distinction between chat and user more refined where possible

Point 5: If you run that code and have logging enabled, you'll see an exception `TypeError: 'mappingproxy' object does not support item assignmen`. As explained in the docs that I linked above, `application.chat_data` is read only, you can only read the values from the dict, not write new values. You can modify the values themself, though. E.g. `context.application.chat_data[chat_id].update({1: 'some text', 2: 'another text'})` will work.

**Reply by avfawkes** [reply]
@Bibo-Joshi There was my stupid bug. I missed that add `t_user_id` to `chat_data` as str, and was trying to find why in inline button callback `chat_data` is empty. There were 2 keys: str with data and int which is empty. Thanks for response, and excuse for wasting you time.

**Reply by Bibo-Joshi** [reply]
Glad you could solve it 👍
